d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java

```diff
@@ -0,0 +1,308 @@
+
+    // subscriber data should be updated to NODE_UPDATE2
+    fixture.getSpiedXdsClientImpl().handleD2NodeResponse(DISCOVERY_RESPONSE_NODE_DATA2);
+    actualData = (XdsClient.NodeUpdate) fixture._nodeResourceSubscriber.getData();
+    verify(fixture._nodeResourceSubscriber).notifyWatcher(eq(fixture._resourceWatcher), eq(NODE_UPDATE2));
+    Assert.assertEquals(actualData.getVersion(), NODE_UPDATE2.getVersion());
+    Assert.assertEquals(actualData.getNodeData(), NODE_UPDATE2.getNodeData());
+  }
+
+  @Test
+  public void testHandleD2URIMapResponseWithData()
+  {
+    XdsClientImplFixture fixture = new XdsClientImplFixture();
+    // subscriber original data is null
+    fixture._nodeResourceSubscriber.setData(null);
+    fixture.getSpiedXdsClientImpl().handleD2URIMapResponse(DISCOVERY_RESPONSE_URI_MAP_DATA1);
+    verify(fixture._xdsClientImpl).handleResourceUpdate(any(), any());
+    verify(fixture._xdsClientImpl).sendAckOrNack(any(), any(), any());
+    verify(fixture._xdsClientImpl).getResourceSubscriberMap(D2_URI_MAP);
+    verify(fixture._uriMapResourceSubscriber).notifyWatcher(eq(fixture._resourceWatcher),
+        eq(D2_URI_MAP_UPDATE_WITH_DATA1));
+    XdsClient.D2URIMapUpdate actualData = (XdsClient.D2URIMapUpdate) fixture._uriMapResourceSubscriber.getData();
+    // subscriber data should be updated to D2_URI_MAP_UPDATE_WITH_DATA1
+    Assert.assertEquals(actualData.getVersion(), D2_URI_MAP_UPDATE_WITH_DATA1.getVersion());
+    Assert.assertEquals(actualData.getURIMap(), D2_URI_MAP_UPDATE_WITH_DATA1.getURIMap());
+    fixture.getSpiedXdsClientImpl().handleD2URIMapResponse(DISCOVERY_RESPONSE_URI_MAP_DATA2);
+    actualData = (XdsClient.D2URIMapUpdate) fixture._uriMapResourceSubscriber.getData();
+    // subscriber data should be updated to D2_URI_MAP_UPDATE_WITH_DATA2
+    verify(fixture._uriMapResourceSubscriber).notifyWatcher(eq(fixture._resourceWatcher),
+        eq(D2_URI_MAP_UPDATE_WITH_DATA2));
+    Assert.assertEquals(actualData.getVersion(), D2_URI_MAP_UPDATE_WITH_DATA2.getVersion());
+    Assert.assertEquals(actualData.getURIMap(), D2_URI_MAP_UPDATE_WITH_DATA2.getURIMap());
+  }
+
+  @Test
+  public void testHandleD2NodeUpdateWithEmptyData()
+  {
+    XdsClientImplFixture fixture = new XdsClientImplFixture();
+    fixture.getSpiedXdsClientImpl().handleD2NodeResponse(DISCOVERY_RESPONSE_WITH_EMPTY_NODE_RESOURCES);
+    verify(fixture._xdsClientImpl).handleResourceUpdate(eq(new HashMap<>()), eq(NODE));
+    verify(fixture._xdsClientImpl).sendAckOrNack(any(), any(), any());
+  }
+
+  @DataProvider(name = "badNodeUpdateTestCases")
+  public Object[][] provideBadNodeDataTestCases()
+  {
+    return new Object[][]{{DISCOVERY_RESPONSE_NODE_RESOURCE_IS_NULL},
+        {DISCOVERY_RESPONSE_NODE_NULL_DATA_IN_RESOURCE_FILED},};
+  }
+
+  @Test(dataProvider = "badNodeUpdateTestCases")
+  public void testHandleD2NodeUpdateWithBadData(DiscoveryResponseData badData)
```

Comment:
all the bad data tests below are missing a case where the initial data in the subscriber is null.